### PR TITLE
[MRG] ENH: Add option to customize spike histogram colors

### PR DESCRIPTION
**Pull request by ntolley**
As title states.

### CODE REVIEW

File: hnn_core/cell_response.py [hunk 1]
```diff
@@ -370,6 +370,13 @@ def plot_spikes_hist(self, trial_idx=None, ax=None, spike_types=None,
             Valid strings also include leading characters of spike types
 
             | Ex: ``'ev'`` is equivalent to ``['evdist', 'evprox']``
+        color : str | list of str | None
+            String input defining color of histograms plotted. If list of str
+            provided, histograms for each cell type will be plotted by cycling
+            through colors in the list.
+
+            | Ex: ``'r'``, ``['r', 'g']``
+        If None, default color cycle used.
```
Comment (jasmainak) [inline hnn_core/cell_response.py:379]: how do we know which color corresponds to which cell type?

Comment (ntolley) [reply hnn_core/cell_response.py:379]: The legend will indicate the cell type in the plot, but in the code it can be redefined.

File: hnn_core/cell_response.py [hunk 2]
```diff
@@ -370,6 +370,13 @@ def plot_spikes_hist(self, trial_idx=None, ax=None, spike_types=None,
             Valid strings also include leading characters of spike types
 
             | Ex: ``'ev'`` is equivalent to ``['evdist', 'evprox']``
+        color : str | list of str | None
+            String input defining color of histograms plotted. If list of str
+            provided, histograms for each cell type will be plotted by cycling
```
Comment (rythorpe) [inline hnn_core/cell_response.py:375]:
```suggestion
            provided, histograms for each spike type will be plotted by cycling
```
Just to be consistent with our use of the `spike_types` argument.

File: hnn_core/viz.py [hunk 3]
```diff
@@ -353,6 +353,13 @@ def plot_spikes_hist(cell_response, trial_idx=None, ax=None, spike_types=None,
         Valid strings also include leading characters of spike types
 
         | Ex: ``'ev'`` is equivalent to ``['evdist', 'evprox']``
+    color : str | list of str | None
+        String input defining color of histograms plotted. If list of str
+        provided, histograms for each cell type will be plotted by cycling
```
Comment (rythorpe) [inline hnn_core/viz.py:358]:
```suggestion
        provided, histograms for each spike type will be plotted by cycling
```
Same here.

**Comment by rythorpe** [comment]
Aside from my docstring nitpick, LGTM!

File: hnn_core/cell_response.py (hunk 1 above)
Comment (jasmainak) [reply hnn_core/cell_response.py:379]: I was thinking a dict might be more useful?

```py
dict(evprox='r', evdist='g')
```

Comment (rythorpe) [reply hnn_core/cell_response.py:379]: Defining a color sequence that gets cycled over is pretty standard for matplotlib and other high-level plotting libraries (e.g., seaborn) is it not? I'm not opposed to using a dictionary, but it's not as much of a lightweight solution as what is implemented now.

Something I just thought of, do we want it to also accept the RGBA color specification as well?

Comment (jasmainak) [reply hnn_core/cell_response.py:379]: since `net.gid_dict` is a dictionary, it means that the cell types won't always necessarily be in the same order before Python 3.6? Maybe I'm splitting at hairs here, but we should avoid sources of bugs

Comment (rythorpe) [reply hnn_core/cell_response.py:379]: I don't think that would be a bug. This function currently allows users to specify the cell types, drive types, or grouping of types they wish to plot. If the user makes no underlying assumptions about the order in which the various histograms will be plotted, they can rely on the default behavior which is to plot all cell types based on `net.gid_dict`. If we allow a color sequence to be entered, they could also theoretically use a matplotlib colormap!

Comment (ntolley) [reply hnn_core/cell_response.py:379]: Would it be a bug that we don't guarantee plots having the same color if python versions change? Also that means that our current code also doesn't have that assurance, so we're not losing anything necessarily

**Comment by jasmainak** [comment]
okay let's move on if it's good by @chenghuzi !

**Comment by chenghuzi** [comment]
> Sounds good! @chenghuzi just to double check is this PR sufficient for your histogram plots in the GUI?

LGTM! I still need to manage the color myself outside of it, but it’s doable.

**Comment by jasmainak** [comment]
> I still need to manage the color myself outside of it, but it’s doable.

Can you explain a bit more?

**Comment by chenghuzi** [comment]
> > I still need to manage the color myself outside of it, but it’s doable.
> 
> Can you explain a bit more?

So in this case for a single figure, if the user calls the histogram, for example 3 times, the result will be different if we use the `next` iterator to get colors as it's stateful. But we can wrap that operation into the default figure creation function  to make it appear as an all-in-one solution.

**Comment by jasmainak** [comment]
is the problem that `color` is used as an iterator to cycle rather than an iterator that has the same length as the number of drives?

**Comment by ntolley** [comment]
@rythorpe I implemented that better logic/system to aggregate the spike times in a single list. Also added a bunch of tests to check the edge cases for the dictionary input

**Comment by jasmainak** [comment]
@ntolley tests are failing ...

File: hnn_core/cell_response.py [hunk 4]
```diff
@@ -370,6 +370,13 @@ def plot_spikes_hist(self, trial_idx=None, ax=None, spike_types=None,
             Valid strings also include leading characters of spike types
 
             | Ex: ``'ev'`` is equivalent to ``['evdist', 'evprox']``
+        color : str | list of str | None
```
Comment (jasmainak) [inline hnn_core/cell_response.py:373]: does this documentation need to be updated to match the other file?

File: hnn_core/tests/test_viz.py [hunk 5]
```diff
@@ -161,5 +167,22 @@ def test_dipole_visualization():
         net.cell_response.plot_spikes_hist(trial_idx='blah')
     net.cell_response.plot_spikes_hist(trial_idx=0, show=False)
     net.cell_response.plot_spikes_hist(trial_idx=[0, 1], show=False)
+    net.cell_response.plot_spikes_hist(color='r')
+    net.cell_response.plot_spikes_hist(color=['C0', 'C1'])
+    net.cell_response.plot_spikes_hist(color={'beta_prox': 'r',
+                                              'beta_dist': 'g'})
```
Comment (jasmainak) [inline hnn_core/tests/test_viz.py:173]: Do we have a test that checks that it works fine when there are two drives defined to belong to the same grouping in `spike_types` ?

**Comment by jasmainak** [comment]
@chenghuzi does this work for you now with the updates?

File: hnn_core/tests/test_viz.py [hunk 6]
```diff
@@ -161,5 +167,22 @@ def test_dipole_visualization():
...
+
+    with pytest.raises(TypeError, match="color must be an instance of"):
+        net.cell_response.plot_spikes_hist(color=123)
+    with pytest.raises(ValueError):
+        net.cell_response.plot_spikes_hist(color='z')
```
Comment (ntolley) [inline hnn_core/tests/test_viz.py:178]: the warning text changed between versions of matplotlib so there is no match here

File: hnn_core/tests/test_viz.py (hunk 5 above)
Comment (ntolley) [reply hnn_core/tests/test_viz.py:173]: Added!

File: hnn_core/viz.py [hunk 7]
```diff
@@ -353,6 +353,20 @@ def plot_spikes_hist(cell_response, trial_idx=None, ax=None, spike_types=None,
         Valid strings also include leading characters of spike types
 
         | Ex: ``'ev'`` is equivalent to ``['evdist', 'evprox']``
+    color : str | dict | None
+        Input defining colors of plotted histograms. If str, all
+        histograms plotted with same color. If list of str provided,
```
[the remark this answers is not in the document]
Comment (ntolley) [reply hnn_core/viz.py:358]: fixed docs to match

**Review by rythorpe — APPROVED** [review]

**Review by jasmainak — APPROVED** [review]
@chenghuzi merge button is yours if you are able to use it in the GUI :)

**Comment by chenghuzi** [comment]
> 

LGTM!

**Comment by chenghuzi** [comment]
> @chenghuzi merge button is yours if you are able to use it in the GUI :)

Resolved the what's new conflict and now seems everything is good!

**Comment by jasmainak** [comment]
@ntolley can you fix the commit history to remove merge commits and resolve the conflicts ?

**Comment by ntolley** [comment]
@jasmainak all done! Don't think there are any merge commits? It was just a conflict in whats_new.rst

**Comment by rythorpe** [comment]
I'm going to move forward with the rebase and merge. Thanks @ntolley!

**Comment by jasmainak** [comment]
amazing thanks @ntolley ! 🥳
